interp/iostream.go

```diff
@@ -0,0 +1,209 @@
+package interp
+
+// I/O streams are interfaces which allow file redirects and command pipelines to be treated
+// equivalently.
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"os/exec"
+	"syscall"
+)
+
+const (
+	notClosedExitCode = -127.0
+)
+
+var (
+	doubleCloseError = errors.New("close: stream already closed")
+)
+
+// firstError returns the first non-nill error or nil if all errors are nil.
+func firstError(errs ...error) error {
+	for _, err := range errs {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Close the cmd and convert the error result into the result returned from goawk builtin functions.
+// A nil error is returned if that error describes a non-zero exit status or an unhandled signal.
+// Any other type of error returns -1.0 and err.
+//
+// The result mimicks gawk for expected child process errors:
+// 1. Returns the exit status of the child process and nil error on normal process exit.
+// 2. Returns 256 + signal on unhandled signal exit.
+// 3. Returns 512 + signal on unhandled signal exit which caused a core dump.
+func waitExitCode(cmd *exec.Cmd) (float64, error) {
+	err := cmd.Wait()
+	if err == nil {
+		return 0.0, nil
+	}
+	ee, ok := err.(*exec.ExitError)
+	if !ok {
+		// Wait() returned an io error.
+		return -1.0, err
+	}
+	status, ok := ee.ProcessState.Sys().(syscall.WaitStatus)
+	if !ok {
+		// Maybe not all platforms support WaitStatus?
+		return -1.0, err
+	}
+	switch {
+	case status.CoreDump():
+		return 512.0 + float64(status.Signal()), nil
+	case status.Signaled():
+		return 256.0 + float64(status.Signal()), nil
+	case status.Exited():
+		return float64(status.ExitStatus()), nil
+	default:
+		return -1.0, err
+	}
+}
+
+type inputStream interface {
+	io.ReadCloser
+	ExitCode() float64
+}
+
+type outputStream interface {
+	io.WriteCloser
+	Flush() error
+	ExitCode() float64
+}
+
+type outFileStream struct {
+	*bufio.Writer
+	closer   io.Closer
+	exitCode float64
+}
+
+func newOutFileStream(wc io.WriteCloser, size int) outputStream {
+	b := bufio.NewWriterSize(wc, size)
+	return &outFileStream{b, wc, notClosedExitCode}
+}
+
+func (s *outFileStream) Close() error {
+	if s.exitCode != notClosedExitCode {
```

Comment:
I would prefer this were a separate `closed bool` on the type, rather than trying to stuff it into the `exitCode` field. That would avoid the magic `notClosedExitCode` constant, and make this test:

```go
if s.closed {
    return doubleCloseError
}
s.closed = true
```

Reply:
Ok. It seems redundant to me but that's fine. Notice that the only way it could be a negative number, or that specific exit code, is if `exitCode` is initialized with that and never overwritten. But sure.